python/dgl/dataloading/negative_sampler.py

```diff
@@ -65,3 +66,67 @@ def _generate(self, g, eids, canonical_etype):
         src = F.repeat(src, self.k, 0)
         dst = F.randint(shape, dtype, ctx, 0, g.number_of_nodes(vtype))
         return src, dst
+
+# Alias
+Uniform = PerSourceUniform
+
+class GlobalUniform(_BaseNegativeSampler):
+    """Negative sampler that randomly chooses negative source-destination pairs according
+    to a uniform distribution.
+
+    For each edge ``(u, v)`` of type ``(srctype, etype, dsttype)``, DGL generates
+    :attr:`k` pairs of negative edges ``(u', v')``, where ``u'`` is chosen uniformly from
+    all the nodes of type ``srctype`` and ``v'`` is chosen uniformly from all the nodes
+    of type ``dsttype``.  The resulting edges will also have type
+    ``(srctype, etype, dsttype)``.
+
+    Parameters
+    ----------
+    k : int
+        The number of negative examples to generate per minibatch.
+    exclude_self_loops : bool, optional
+        Whether to exclude self-loops from negative examples.  (Default: True)
+    unique : bool, optional
+        Whether to sample unique negative examples.  Setting it to False will make things
+        faster.  (Default: True)
+    num_trials : int, optional
+        The number of rejection sampling trials.
+
+        Increasing it will increase the likelihood of getting :attr:`k` negative examples,
+        but will also take more time if the graph is dense.
+    redundancy : float, optional
+        Indicates how much more negative examples to actually generate during rejection sampling
+        before finding the unique pairs.
+
+        Increasing it will increase the likelihood of getting :attr:`k` negative examples,
+        but will also take more time and memory.
```

Comment:
Is it a standard way to expose both `num_trials` and `redundancy`? They are hard to understand if users are not familiar with the implementation.

Reply:
Perhaps we could calculate "redundancy" based on the sparsity of the graph and hard-coding some probabilistic measure?  For example, the following function calculates the number of node pairs required given the desired "k_hat" number of negative samples:

```python
import numpy as np
from numpy.polynomial import polynomial

def calc_draw(k_hat, num_edges, num_nodes, r=3):
    """
    Args:
    ---------
    k_hat : int
        Number of desired samples
    num_edges : int
        Number of edges in the graph
    num_nodes : int
        Number of nodes in the graph
    r : float
        Number of standard deviations
        
    Calculates the number of samples required based on a lower-bound
    of the expected number of negative samples, based on N draws from 
    a binomial distribution.  Solves the following equation for N:
    
    k_hat = N*p_k - r * np.sqrt(N*p_k*(1-p_k))
        
    where p_k is the probability that a node pairing is a negative edge 
    and r is the number of standard deviations to construct the lower bound
    """
    p_m = num_edges / (num_nodes*(num_nodes-1))
    p_k = 1 - p_m
    
    a = p_k**2
    b = -p_k*(2*k_hat+r**2*p_m)
    c = k_hat**2
    
    poly = polynomial.Polynomial([c, b, a])
    N = poly.roots()[-1]
    redundancy = N / k_hat - 1.
    return redundancy
```

For example, using 3 standard deviations to give a probability >99.9% and the graph stats from DDI (the most dense link prediction graph on OGB by far @ 14% density), I get a redundancy of 21% for a batch size of 1024.  The second most dense graph (biokg @ 0.1% density) gets a redundancy of 0.4%.  

The good news is that this is relatively insensitive to the `r` parameter above, since the variance scales as `sqrt(N*p_k*(1-p_k))`.  For DDI, the redundancies for 1, 3 and 5 standard deviations are 18%, 21%, and 24% respectively.  

Maybe we just set it to something like 5 and issue a warning if something is strange, like a redundancy > 1...etc?  

For the number of trails, perhaps a similar analysis can be done based on the combinatorics.  For example, the number of node combinations in DDI is 9.1M, and if 85% of those are negative edges, that's 7.8M.  For a batch size of 1024, there's around 0.01% chance you would have a duplicate.  In that case, maybe we just get the redundancy right and remove the num_trials option?

Reply:
Yes.  If we can determine the redundancy intelligently we can remove both arguments actually.  I would probably hardcode the number of trials to 3 or something.

Reply:
Changed the implementation of `global_uniform_negative_sampling` to include @zjost 's suggestion.  Now it only has a single redundancy argument which defaults to None (where the redundancy is inferred from the density of the graph).  The number of trials is fixed to 3.